feeadjuster/feeadjuster.py

```diff
@@ -0,0 +1,128 @@
+                plugin.log("Could not adjust fees for channel {}: '{}'"
+                           .format(scid, e), level="warn")
+
+
+def get_chan(plugin: Plugin, scid: str):
+    for peer in plugin.rpc.listpeers()["peers"]:
+        if len(peer["channels"]) == 0:
+            continue
+        chan = peer["channels"][0]
+        if "short_channel_id" not in chan:
+            continue
+        if chan["short_channel_id"] == scid:
+            return chan
+
+
+def maybe_add_new_balances(plugin: Plugin, scids: list):
+    for scid in scids:
+        if scid not in plugin.adj_balances:
+            # At this point we could call listchannels and pass the scid as
+            # argument, and deduce the peer id (!our_id). But -as unlikely as
+            # it is- we may not find it in our gossip (eg some corruption of
+            # the gossip_store occured just before the forwarding event).
+            # However, it MUST be present in a listpeers() entry.
+            chan = get_chan(plugin, scid)
+            assert chan is not None
+
+            plugin.adj_balances[scid] = {
+                "our": int(chan["to_us_msat"]),
+                "total": int(chan["total_msat"])
+            }
+
+
+@plugin.subscribe("forward_event")
+def forward_event(plugin: Plugin, forward_event: dict, **kwargs):
+    if forward_event["status"] == "settled":
+        in_scid = forward_event["in_channel"]
+        out_scid = forward_event["out_channel"]
+        maybe_add_new_balances(plugin, [in_scid, out_scid])
+
+        plugin.adj_balances[in_scid]["our"] += forward_event["in_msatoshi"]
+        plugin.adj_balances[out_scid]["our"] -= forward_event["out_msatoshi"]
+        try:
+            # Pseudo-randomly add some hysterisis to the update
+            if not plugin.deactivate_fuzz and random.randint(0, 9) == 9:
+                time.sleep(random.randint(0, 5))
+            maybe_adjust_fees(plugin, [in_scid, out_scid])
+        except Exception as e:
+            plugin.log("Adjusting fees: " + str(e), level="error")
+
+
+@plugin.init()
+def init(options: dict, configuration: dict, plugin: Plugin, **kwargs):
+    plugin.our_node_id = plugin.rpc.getinfo()["id"]
+    plugin.deactivate_fuzz = options.get("feeadjuster-deactivate-fuzz", False)
+    config = plugin.rpc.listconfigs()
+    plugin.adj_basefee = config["fee-base"]
+    plugin.adj_ppmfee = config["fee-per-satoshi"]
+
+    for peer in plugin.rpc.listpeers()["peers"]:
+        if len(peer["channels"]) == 0:
+            continue
+        chan = peer["channels"][0]
+        if "short_channel_id" not in chan:
+            continue
+        if chan["state"] != "CHANNELD_NORMAL":
+            continue
+        # We don't store it as a percentage, to be able to update in on
+        # forward without calling listpeers()
+        plugin.adj_balances[chan["short_channel_id"]] = {
+                "our": int(chan["to_us_msat"]),
+                "total": int(chan["total_msat"])
+        }
```

Comment:
Am I right that this block is not required as the channels are build up on demand anyway within `forward_event` and `maybe_add_new_balances`?

Reply:
Just checked: pytests are still green if this block is commented out.

Reply:
You are right, it's not required. The rationale was to avoid updating the whole list on first forward but at startup instead. Now i'm not sure it makes much sense

Reply:
Ok, removed it for the sake of me not able to justify it clearly.